source/server/listener_impl.cc

```diff
@@ -232,30 +252,92 @@ ListenerImpl::ListenerImpl(const envoy::config::listener::v3::Listener& config,
           listener_init_target_.ready();
         }
       }) {
-  Network::Address::SocketType socket_type =
-      Network::Utility::protobufAddressSocketType(config.address());
-  if (PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, transparent, false)) {
-    addListenSocketOptions(Network::SocketOptionFactory::buildIpTransparentOptions());
+  buildAccessLog();
+  auto socket_type = Network::Utility::protobufAddressSocketType(config.address());
+  this->buildListenSocketOptions(socket_type)
+      ->buildUdpListenerFactory(socket_type, concurrency)
+      ->createListenerFilterFactories(socket_type);
+  this->validateFilterChains(socket_type)->buildFilterChains();
+  if (socket_type == Network::Address::SocketType::Datagram) {
+    return;
   }
-  if (PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, freebind, false)) {
-    addListenSocketOptions(Network::SocketOptionFactory::buildIpFreebindOptions());
+  this->buildSocketOptions()
+      ->buildOriginalDstListenerFilter()
+      ->buildProxyProtocolListenerFilter()
+      ->buildTlsInspectorListenerFilter();
+  if (!workers_started_) {
+    // Initialize dynamic_init_manager_ from Server's init manager if it's not initialized.
+    // NOTE: listener_init_target_ should be added to parent's initManager at the end of the
+    // listener constructor so that this listener's children entities could register their targets
+    // with their parent's initManager.
+    parent_.server_.initManager().add(listener_init_target_);
   }
-  if (config.reuse_port()) {
-    addListenSocketOptions(Network::SocketOptionFactory::buildReusePortOptions());
-  } else if (socket_type == Network::Address::SocketType::Datagram && concurrency > 1) {
-    ENVOY_LOG(warn, "Listening on UDP without SO_REUSEPORT socket option may result to unstable "
-                    "packet proxying. Consider configuring the reuse_port listener option.");
+}
+
+ListenerImpl::ListenerImpl(const ListenerImpl& origin,
+                           const envoy::config::listener::v3::Listener& config,
+                           const std::string& version_info, ListenerManagerImpl& parent,
+                           const std::string& name, bool added_via_api, bool workers_started,
+                           uint64_t hash, uint32_t concurrency)
+    : parent_(parent), address_(origin.address_),
```

Comment:
There is a lot of copying in here. Is it possible to have a shared constructor that is called from both the update and new path? This might be a moot point depending on the outcome of my comment around doing an in place upgrade while preserving the tag.

Reply:
ping on this. It seems highly likely that there can be a private shared constructor used by both current constructors? You might be able to get rid of the utility function changes that way and greatly reduce the amount of changes in this file.

Reply:
This is a non-trivial work:
Exception safe, some field initialize order must be respected. 

The two constructor is the best shape, but moving to shared constructor is not a low hanging fruit.
I prefer to explore reuse ListenerImpl before looking into the shared constructor :)